rendezvous/README.md

```diff
@@ -0,0 +1,173 @@
+discovery within application specific namespaces. Peers connect to the
+rendezvous point and register their presence in one or more
+namespaces.
+
+Peers registered with the rendezvous point can be discovered by other
+nodes by querying the rendezvous point. The query specifies the
+namespace for limiting application scope and optionally a maximum
+number of peers to return. The namespace can be omitted in the query,
+which asks for all peers registered to the rendezvous point.
+
+The query can also specify a timestamp, obtained from the response to
+a previous query, such that only peers registered since the timestamp
+will be returned. This allows peers to progressively refresh their
+network view without overhead, which greatly simplifies real time
+discovery.
+
+### Registration Lifetime
+
+Registration lifetime is controlled by an optional TTL parameter in
+the `REGISTER` message.  If a TTL is specified, then the registration
+persists until the TTL expires.  If no TTL was specified, then the
+registration persists while the issuing peer is still connected to the
+rendezvous point.
+
+Peers can refresh their registrations at any time with a new
+`REGISTER` message; the TTL of the new message supersedes previous
+registrations. Peers can also cancel existing registrations at any
+time with an explicit `UNREGISTER` message.
+
+### Interaction
+
+Clients `A` and `B` connect to the rendezvous point `R` and register for namespace
+`my-app` with a `REGISTER` message:
+
+```
+A -> R: REGISTER{my-app, {QmA, AddrA}}
+B -> R: REGISTER{my-app, {QmB, AddrB}}
+```
+
+Client `C` connects and registers for namespace `another-app`:
+```
+C -> R: REGISTER{another-app, {QmC, AddrC}}
+```
+
+Another client `D` can discover peers in `my-app` by sending a `DISCOVER` message; the
+rendezvous point responds with the list of current peer reigstrations.
+```
+D -> R: DISCOVER{ns: my-app}
+R -> D: {[REGISTER{my-app, {QmA, Addr}}
+          REGISTER{my-app, {QmB, Addr}}],
+         t1}
+```
+
+If `D` wants to discover all peers registered with `R`, then it can omit the namespace
+in the query:
+```
+D -> R: DISCOVER{}
+R -> D: {[REGISTER{my-app, {QmA, Addr}}
+          REGISTER{my-app, {QmB, Addr}}
+          REGISTER{another-app, {QmC, AddrC}}],
+         t2}
+```
+
+If `D` wants to progressively poll for real time discovery, it can use
+the timestamp obtained from a previous response in order to only ask
+for new registrations.
+
+So here we consider a new client `E` registering after `t1`, and a subsequent
+query that discovers just that peer by including the timestamp:
+
+```
+E -> R: REGISTER{my-app, {QmE, AddrE}}
+D -> R: DISCOVER{ns: my-app, since: t1}
+R -> D: {[REGISTER{my-app, {QmE, AddrE}}],
+         t3}
+```
+
+### Protobuf
+
+```protobuf
+message Message {
+  enum MessageType {
+    REGISTER = 0;
+    UNREGISTER = 1;
+    DISCOVER = 2;
+    DISCOVER_RESPONSE = 3;
+  }
+
+  message PeerInfo {
+    optional string id = 1;
```

Comment:
If there is no crypto-challenge for the id the only way to prove ownership is the SECIO challenge, which is only done for the current peer-id. Therefore this field seems useless to me...
Why not just drop the PeerInfo msg and use repeated addrs on register?

Reply:
> You also are missing the potential use case of rendezvous points sharing registrations (say in federation of daemons, using pubsub, etc).

Then the data must be somehow authenticated, right?

Reply:
You can authenticate the sharing peer instead of the data.
Authenticating the registration itself would require a signature in them.

Reply:
> You can authenticate the sharing peer instead of the data.

If the data isn't authenticated, wouldn't that allow for replay attacks?

Reply:
Sure, but you'd get that too if you were authenticating with a nonce.

I don't think that replay attacks are a big concern in this context, as the sharing peers can establish a trust model with each other.

Note that signatures along are not sufficient to prevent replay attacks in a shared setting either, and trying to add a signature timestamp gets complicated quick.

Reply:
Ok